Fix pagination issue

modhurita opened the pull request:
Hi @qubixes :

I fixed the problem of all artworks not being scrolled through with the right-arrow. If this looks good, can you please accept the pull request?

Thanks!

qubixes commented on artscraper/find_artworks.py:
```diff
@@ -209,26 +209,46 @@ def get_artist_works(self):
-        while right_arrow_element.get_attribute('tabindex') is not None:
+        # Find number of artists
+        # Find elements with tag name 'h3'
+        items_elements = parent_element.find_elements('tag name', 'h3')
```
What happens if somehow you can't find any match? Then total_num_artworks will be uninitialized?

modhurita replied:
I have addressed this now - it should now work whether or not total_num_elements exists.

qubixes replied:
I have read through the code, but I'm not quite sure. I think you need to need to set total_num_elements to some value, otherwise you'll still get a NameError later, since you didn't assign it anything, or am I missing something?

qubixes commented on artscraper/find_artworks.py:
```diff
@@ -209,26 +209,46 @@ def get_artist_works(self):
+        # Initialize number of artworks
+        num_artworks = 0
+
+        while num_artworks < total_num_artworks:
```
What if the num_artworks never reaches total_num_artworks? It will then simply loop endlessly right? Would be good in that case to have a maximum number of tries (with the same results).

modhurita replied:
I now make it try thrice before breaking out of the loop.

qubixes reviewed (changes requested):
Nice @modhurita I only have two comments, both directed at when something might go wrong/is incorrect on the google site.

I haven't tried to run the code myself, I trust you have tried it yourself?

Either way, looks good otherwise!

modhurita commented:
Hi @qubixes :

Thanks for your feedback and for pointing out possible problems. I have made changes to address your concerns. Does it look ok now?

qubixes commented on artscraper/find_artworks.py:
```diff
@@ -209,26 +209,77 @@ def get_artist_works(self):
+        # Initialize count of number of iterations for which the number of artworks remains the same
+        count = 0
+
+        while True:
```
I think there we're still in the possible danger zone of infinite loops. If the right arrow button can be clicked, but won't add new elements to the list, then it will continue to loop.

There also seems to be some duplication in this function. Below a pseudo-code suggestion to make this part of the code easier to read.

```python
art_list = _get_art_list(...)
while (len(art_list) < tot_art)) and not (tot_art == 0 and n_try > 3):
    if _button_clickable():
        wait()
        click_button()
    art_list = _get_art_list()
    if len(art_list) == prev_len:
       n_try += 1
    else:
       n_try = 0
```

It might also be helpful to put theses part of the code in small functions (you can define them inline). Please check if the logic above is correct if you decide to implement it that way!

qubixes reviewed (changes requested):
Thanks @modhurita, I think there are still some issues. Let me know if you need any help if it's unclear (I have time on Thursday).

modhurita commented:
Hi @qubixes :

I have addressed your concerns and rewritten the code in the manner you suggested. Does it look ok now? Please let me know if there are still some possibilities I have overlooked, or if you want the code improved further.

I don't quite understand why you are checking for whether the total number of artworks is 0 - is that a possibility? I thought each artist had at least 3 artworks.

qubixes commented:
@modhurita The logic seems sound to me now in that loop. The total number of artworks to be zero would be in case it can't find the number, in the previous part. But in the way you have programmed it, it is not necessary indeed. Instead it could be `while len(list) < total *or* n_try > 3:`. Can you look at the comment on getting the total number of artworks? If those are all correct, then I think we can merge soon.

modhurita commented:
Hi @qubixes 

I have made the latest change we discussed - I removed the check for the total number of artworks being zero. Does the code look good now?

qubixes reviewed (approved):
Approved!
